### ISSUE TITLE
DPOTrainer is slower compared to custom training loop

### ISSUE THREAD

**Issue by abaheti95** [huggingface/trl#729]
Hello,
Thank you for sharing DPOTrainer in your library. I am using it to finetune [timdettmers/qlora-hh-rlhf-7b](https://huggingface.co/timdettmers/qlora-hh-rlhf-7b) PeftModel on [Anthropic/hh-rlhf](https://huggingface.co/datasets/Anthropic/hh-rlhf). I am following [dpo_llama2.py](https://github.com/huggingface/trl/blob/main/examples/research_projects/stack_llama_2/scripts/dpo_llama2.py) example reference to prepare my training script.

The training speed is also significantly slower than a simple training loop I wrote for a supervised finetuning setup with `PeftModel`. That training loop takes about `1.2s` for a batch size of `4` i.e. `9.6s` for `32` instances. However, `32` instances in DPOTrainer with the same model are taking about `98.06s`.  

--------------------------

For reference, this is how I initialize the model
```python
base_model = AutoModelForCausalLM.from_pretrained(
        script_args.model_name_or_path,
        load_in_4bit=True,
        torch_dtype=torch.bfloat16,
        device_map={"": 0},
        quantization_config=BitsAndBytesConfig(
            load_in_4bit=True,
            bnb_4bit_compute_dtype=torch.bfloat16,
            bnb_4bit_use_double_quant=True,
            bnb_4bit_quant_type='nf4'
        ),
    )
model = PeftModel.from_pretrained(base_model, script_args.adapter_path, is_trainable=True).to(device)
```
Here are my library versions
```bash
peft==0.5.0
trl=0.7.1
bitsandbytes==0.41.1
transformers==4.32.1
accelerate==0.22.0
```

**Comment by lvwerra** [comment]
Is your own loop also using PEFT?

cc @kashif and @younesbelkada

**Comment by abaheti95** [comment]
Yes, I am using the Peft models for training. I also implemented a custom training loop on my own which is taking about half the time.

**Comment by younesbelkada** [comment]
Hi @abaheti95 
double quantization is slower than fine-tuning without it, do you also use `bnb_4bit_use_double_quant=True` on your SFT script? 
You can also try to set the compute dtype to `float16` instead of `bfloat16`
To make it even faster you can also use `BetterTransformer` API and call 

```python
base_model.to_bettertransformer()
```

Before:

```python
model = PeftModel.from_pretrained(base_model, script_args.adapter_path, is_trainable=True).to(device)
```

**Comment by abaheti95** [comment]
Wow! Thank you for the tips. I was unaware of many of these tricks. I was following the setup that I thought was the least memory-intensive as per blogs and papers. I will look into the documentation for each of these suggestions and try to make it more efficient. 

Mainly I want to make both training and inference faster while consuming less memory. 
I am using llama7b instead of llama2 because the pretrained peft model is for llama7b. 
> do you also use `bnb_4bit_use_double_quant=True` on your SFT script?
I was following the setting from the work who pretrained the adapter: https://github.com/artidoro/qlora/blob/main/qlora.py#L324

Another question: Can we load two adapters over the same base model in which one is trainable (main model) and another is untrainable (dpo - reference model)?

**Comment by abaheti95** [comment]
There seems to be slight differences in greedy decoding when changing dtype `float16` instead of `bfloat16`

**Comment by younesbelkada** [comment]
Thanks for your message!

> I was following the setting from the work who pretrained the adapter: https://github.com/artidoro/qlora/blob/main/qlora.py#L324

I see, note that this will make training slower, if you have enough GPU memory I would advise to turn that flag off ! Use it only if it cause OOM issues without that flag

> Another question: Can we load two adapters over the same base model in which one is trainable (main model) and another is untrainable (dpo - reference model)?

In that case you don't need two adapters, one is sufficient, to use the reference model you would simply need to switch off the main adapter by calling `model.disable_adapters()`, you can read more about it here: https://huggingface.co/blog/trl-peft . I think we do have that support currently on DPO Trainer

> There seems to be slight differences in greedy decoding when changing dtype float16 instead of bfloat16. What could be the reason?

This is expected, there are small numerical differences that can be introduced when using float16 vs bfloat16, as long as the output are not gibberish those differences are expected I would say

**Comment by abaheti95** [comment]
@younesbelkada Sorry for the late response. I was busy with another project. Regarding your suggestion about deactivating the adapters: I wanted to use a pretrained adapter as the starting reference policy and continue finetuning it further with a different loss function. I don't think deactivating functionality will be useful in this case. Is there a way to load two adapters simultaneously (one for training and one for reference?)

**Comment by younesbelkada** [comment]
Hmm I see, perhaps what you can do is merge the first adapter into the active model to create the reference policy and attach the new adapter on top of the merged model

https://huggingface.co/docs/peft/conceptual_guides/lora#merge-lora-weights-into-the-base-model

**Comment by abaheti95** [comment]
I thought of this, but one issue is that after merging the base model changes. The subsequent adapter that is trained will no longer be compatible with the original model (LLaMA-7B) in this case.

**Comment by kashif** [comment]
@abaheti95 So at the moment there is no support for different adaptors for your use case... if you can hash out how one could potentially use other adaptors I can try to add it... one thought is to check if the ref_model is an adaptor in which case it should load that to calculate the log-probs?

**Comment by abaheti95** [comment]
In my scenario, the ref model is a QLoRA adapter on top of LLaMA-7B ([timdettmers/qlora-hh-rlhf-7b](https://huggingface.co/timdettmers/qlora-hh-rlhf-7b)).
My solution was to simply load two separate LLaMA-7B base models with two different adapters (one reference and the other with gradients).
